cfgov/filing_instruction_guide/wagtail_hooks.py

```diff
@@ -0,0 +1,12 @@
+from wagtail.core import hooks
+
+from filing_instruction_guide.models import FIGContentPage
+
+
+def assign_content_section_ids(request, page):
+    if page.specific_class == FIGContentPage:
+        page.assign_section_ids()
+
+
+hooks.register("after_create_page", assign_content_section_ids)
+hooks.register("after_edit_page", assign_content_section_ids)
```

Comment:
As an alternative to using Wagtail hooks here, you could take a look at instead using a custom Page model form, which, per the Wagtail documentation [here](https://docs.wagtail.org/en/v2.15.5/advanced_topics/customisation/page_editing_interface.html#customising-generated-forms) can be used "to automatically generate field content". In the code example linked there, the form's `save()` method sets page fields based on the content of other page fields; it seems like this could also be used to modify a field's content as you are doing here.

On a custom form you can also use `clean()` to validate page content. We use a custom Page model form to perform some extra validation checks on JobListingPages [here](https://github.com/cfpb/consumerfinance.gov/blob/f02731f3a44ec3e9ab030e75dc0b4ffa655c2e6b/cfgov/jobmanager/models/pages.py#L53-L74), which I believe is our only current use of a custom `WagtailAdminPageForm`.

Under the hooks-based approach, if a user creates a new page and then previews it without saving, the hooks never get called, and so the section numbers don't appear. Similarly, if a page is published and then a user makes an section-impacting edit and then previews without saving, the outdated previously-saved numbers appear instead of the new ones.

Additionally, the page is first saved based on what happens in the admin and then saved again in the hook. What I like about the custom form approach is that it allows custom page data validation/customization logic to run with the rest of the code happening in the admin, which means that only a single save is needed (or no saves, when previewing).

If the current implementation meets your needs, that's great, but I thought this might be another option to explore.

Reply:
Thanks for suggesting this! Since I never use the preview workflow, I hadn't noticed that it doesn't update the numbering 😬  . I'll look into the `clean()` form strategy.

Reply:
I just pushed an update using the custom form strategy. That works much better and is more straight forward, good suggestion. Thanks for taking the time to explain it @chosak!